File: Week 4 (10-7)/Bronze/notlast.java
import java.util.*;
import java.io.*;

public class NotLast {

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new FileReader("notlast.in"));
        PrintWriter pw = new PrintWriter(new File("notlast.out"));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int n = Integer.parseInt(st.nextToken());
        String[] names = new String[] { "Bessie", "Elsie", "Daisy", "Gertie", "Annabelle", "Maggie", "Henrietta" };
        int[] cows = new int[7];
        for (int i = 0; i < 7; i++)
            cows[i] = 0;
        for (int i = 0; i < n; i++) {
            st = new StringTokenizer(br.readLine());
            String name = st.nextToken();
            int val = Integer.parseInt(st.nextToken()), index = 0;
            switch (name) {
            case "Bessie":
                index = 0;
                break;
            case "Elsie":
                index = 1;
                break;
            case "Daisy":
                index = 2;
                break;
            case "Gertie":
                index = 3;
                break;
            case "Annabelle":
                index = 4;
                break;
            case "Maggie":
                index = 5;
                break;
            case "Henrietta":
                index = 6;
                break;
            }
            cows[index] += val;
        }
        int minimum = 1000000000;
        for (int i = 0; i < 7; i++) {
            if (cows[i] < minimum) {
                minimum = cows[i];
            }
        }
        int minimum2 = 1000000000, index = 0;
        for (int i = 0; i < 7; i++) {
            if (cows[i] < minimum2 && cows[i] != minimum) {
                index = i;
                minimum2 = cows[i];
            }
        }
        int count = 0;
        for (int i = 0; i < 7; i++) {
            if (cows[i] == minimum2)
                count++;
        }
        pw.println(count == 1 ? names[index] : "Tie");

        br.close();
        pw.close();
    }
}

File: Week 7 (10-28)/Bronze/notlast.cpp
/*
The steps taken to solve the problem are as follows:
- create a hashmap to store the milk output of each cow
- loop through the input and update the milk output to the respective cow
- find the minimum (we need the minimum to figure out the second to last value/cow(s))
- using the minimum, find the second to last cow(s)
  - similar to finding the minimum except you also want to make sure that the value isn't equal to the minimum
*/

#include <bits/stdc++.h>
using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(NULL);
    if (fopen("notlast.in", "r")) {
        freopen("notlast.in", "r", stdin);
        freopen("notlast.out", "w", stdout);
    }

    int n;
    cin >> n;

    unordered_map<string, int> data = {
        {"Bessie", 0},
        {"Elsie", 0},
        {"Daisy", 0},
        {"Gertie", 0},
        {"Annabelle", 0},
        {"Maggie", 0},
        {"Henrietta", 0}
    };
    //Bessie, Elsie, Daisy, Gertie, Annabelle, Maggie, and Henrietta
    for (int i = 0; i < n; i++) {
        string name;
        int change;
        cin >> name >> change;
        data[name] += change;
    }

    int m = 1000000000;
    for (auto el : data) {
        m = min(m, el.second);
    }

    int low = 1000000000;
    vector<string> names;
    for (auto el : data) {
        if (el.second < low && el.second > m) {
            names.clear();
            names.push_back(el.first);
            low = el.second;
        }
        else if (el.second == low) {
            names.push_back(el.first);
        }
    }

    if (names.size() > 1 || names.size() == 0) {
        cout << "Tie\n";
    } else {
        cout << names[0] << "\n";
    }

    return 0;
}